use-timescale/compression/compression-design.md

```diff
@@ -0,0 +1,135 @@
+section covers some of the things you need to take into consideration when
+designing your database for maximum compression effectiveness.
+
+## Array format
+
+TimescaleDB is built on PostgreSQL which is, by nature, a row-based database.
+Because time-series data is accessed in order of time, TimescaleDB converts many
+wide rows of data into a single row of data, called an array form. This means
+that each field of that new, wide row stores an ordered set of data comprising
+the entire column.
+
+For example, if you had a table with data that looked a bit like this:
+
+|Timestamp|Device ID|Status Code|Temperature|
+|-|-|-|-|
+|12:00:01|A|0|70.11|
+|12:00:01|B|0|69.70|
+|12:00:02|A|0|70.12|
+|12:00:02|B|0|69.69|
+|12:00:03|A|0|70.14|
+|12:00:03|B|4|69.70|
+
+You can convert this to a single row in array form, like this:
+
+|Timestamp|Device ID|Status Code|Temperature|
+|-|-|-|-|
+|[12:00:01, 12:00:01, 12:00:02, 12:00:02, 12:00:03, 12:00:03]|[A, B, A, B, A, B]|[0, 0, 0, 0, 0, 4]|[70.11, 69.70, 70.12, 69.69, 70.14, 69.70]|
+
+Even before you compress any data, this format immediately saves storage by
+reducing the per-row overhead. PostgreSQL typically adds around 27 bytes of
+overhead per row. So even without any compression, if our schema above is say 32
+bytes, then 1000 rows of data which previously took about 59 kilobytes
+(`1000 x (32 + 27) ~= 59 kilobytes`), now takes about 32 kilobytes
+(`1000 x 32 + 27 ~= 32 kilobytes`) in this format.
+
+This format arranges the data so that similar data, such as timestamps, device
+IDs, or temperature readings, is stored contiguously. This means that you can
+then use type-specific compression algorithms to compress the data further, and
+each array is separately compressed. For more information about the compression
+methods used, see the [compression methods section][compression-methods].
+
+When the data is in array format, you can perform queries that require a subset
+of the columns very quickly. For example, if you have a query like this one, that
+asks for the average temperature over the past day:
+
+<CodeBlock canCopy={false} showLineNumbers={false} children={`
+SELECT time_bucket(‘1 minute’, timestamp) as minute
+ AVG(temperature)
+FROM table
+WHERE timestamp > now() - interval ‘1 day’
+ORDER BY minute DESC
+GROUP BY minute;
+`} />
+
+The query engine can fetch and decompress only the timestamp and temperature
+columns to efficiently compute and return these results.
+
+Finally, TimescaleDB uses non-inline disk pages to store the compressed arrays.
+This means that the in-row data points to a secondary disk page that stores the
+compressed array, and the actual row in the main table becomes very small,
+because it is now just pointers to the data. When data stored like this is
+queried, only the compressed arrays for the required columns are read from disk,
+further improving performance by reducing disk reads and writes.
+
+## Querying compressed data
```

Comment:
I might need some more concrete examples from you here, if you're able ... ?

Reply:
Here is a very rough blurb of text for the "order-by" example I suggested above.

> ## Ordering data in compressed columns
> 
> A very common way to compress data is to use delta encoding. The default
> algorithm for storing timestamps is the delta-delta algorithm, which will
> compress the data even more, but for this section the delta-encoding is used
> since it is easier to explain. Using delta encoding on the timestamp column will
> store the difference between consecutive values in the column, like this:
> 
> | Timestamp | Device ID | Status Code | Temperature |
> |-----------|-----------|-------------|-------------|
> | 12:00:01  | A         | 0           | 70.11       |
> | 1 second  | A         | 0           | 70.12       |
> | 1 second  | A         | 0           | 70.14       |
> | 12:00:01  | B         | 0           | 69.70       |
> | 1 second  | B         | 0           | 69.69       |
> | 1 second  | B         | 4           | 69.70       |
> 
> Since timestamps are stored at microsecond precision and are represented as a
> 64-bit integer, this would allow the timestamp column to be stored as 64 bits
> for the time and the increments using 20 bits for a total of 26 bytes rather
> than 48 bytes giving a compression of 46%.
> 
> If you looked carefully at the example above, you might have noted that the time
> column was not increasing in time. This can happen if you insert data into the
> hypertable in batches but it creates a problem for the compression
> algorithms. If the rows were ordered in timestamp order instead, you would be
> able to compress the data even more:
> 
> | Timestamp | Device ID | Status Code | Temperature |
> |-----------|-----------|-------------|-------------|
> | 12:00:01  | A         | 0           | 70.11       |
> | 0 seconds | B         | 0           | 69.70       |
> | 1 second  | A         | 0           | 70.12       |
> | 0 seconds | B         | 0           | 69.69       |
> | 1 second  | A         | 0           | 70.14       |
> | 0 seconds | B         | 4           | 69.70       |
> 
> This would compress the timestamps to 64 bits plus 2 times 20 bits, which is a
> whopping 73% compression!
> 
> TimescaleDB will automatically order the rows using the time column, but it also
> allows you to add additional columns to use for ordering using the
> [`compression_orderby` option][compression_orderby].
>

Reply:
Here is a very rough version of the "query compressed data" that brings in a query on the segment-by It is basically using the existing section, but adds a query:

> ## Using segment-by to improve query performance for compressed data 
> 
> The compressed data in TimescaleDB are treated as opaque data by PostgreSQL,
> meaning that it is not possible to look at a compressed batch and know what
> information is in that batch. To be able to figure out if a batch needs to be
> compressed when reading it in a query, TimescaleDB automatically includes more
> information for each batch to improve query performance.
> 
> For order-by columns, TimescaleDB automatically adds additional data that stores
> the minimum and maximum value of that column. This way, the query planner can
> look at the range of timestamps in the compressed column, without having to
> decompress a batch, and determine whether the data in the batch could possibly
> match the query.
> 
> If you, however, want to query based on the device ID, like this, would have to
> decompress the device column to decide if that batch should be part of the
> query.
> 
> <CodeBlock canCopy={false} showLineNumbers={false} children={`
> SELECT time_bucket(‘1 minute’, timestamp) as minute
>  AVG(temperature)
> FROM table
> WHERE timestamp > now() - interval ‘1 day’
>   AND device_id = 'A'
> ORDER BY minute DESC;
> `} />
> 
> If you are often running queries like this, it would be better to not compress
> the device id column at all and instead create batches like this.
> 
> | Device ID | Timestamp                      | Status Code | Temperature           | Min Timestamp | Max Timestamp |
> |-----------|--------------------------------|-------------|-----------------------|---------------|---------------|
> | A         | [12:00:01, 12:00:02, 12:00:03] | [0, 0, 0]   | [70.11, 70.12, 70.14] | 12:00:01      | 12:00:03      |
> | B         | [12:00:01, 12:00:02, 12:00:03] | [0, 0, 0]   | [70.11, 70.12, 70.14] | 12:00:01      | 12:00:03      |
> 
> With the data segmented in this way, a query for device A between a time
> interval becomes quite fast. The query planner can use an index to find those
> rows for device A that contain at least some timestamps corresponding to the
> specified interval, and even a sequential scan is quite fast since evaluating
> device IDs or timestamps does not require decompression. This means the query
> executor only decompresses the timestamp and temperature columns corresponding
> to those selected rows.
> 
> Fortunately, TimescaleDB allow you to provide zero or more segment-by columns
> using the `compression_segmentby` option, which allows you to get good
> performance also for this case.
>